k-distribution/k-tutorial/1_basic/22_proofs/README.md

```diff
@@ -1 +1,444 @@
+
+    rule <k> return I:Int ; ~> _ => I ~> CONT </k>
+         <stack> ListItem(state(CONT, STORE)) => .List ... </stack>
+         <store> _ => STORE </store>
+
+    rule <k> return I:Int ; ~> . => I </k>
+         <stack> .List </stack>
+
+    syntax KItem ::= #makeBindings(Ids, Ints)
+                   | state(continuation: K, store: Map)
+ // ----------------------------------------------------
+    rule <k> #makeBindings(.Ids, .Ints) => . ... </k>
+    rule <k> #makeBindings((I:Id, IDS => IDS), (IN:Int, INTS => INTS)) ... </k>
+         <store> STORE => STORE [ I <- IN ] </store>
+endmodule
+```
+
+Next, compile this example using `kompile lesson-22.k --backend haskell`. If
+your processor is an Apple Silicon processor, add the `--no-haskell-binary`
+flag if the compilation fails.
+
+## 2. Setup: Proof Environment
+
+Next, take the following snippet of K code and save it in `lesson-22-spec.k`.
+This is a skeleton of the proof environment, and we will complete it as the
+lesson progresses.
+
+```
+requires "lesson-22.k"
+
+module LESSON-22-SPEC-SYNTAX
+    imports LESSON-22-SYNTAX
+
+endmodule
+
+module VERIFICATION
+    imports K-EQUAL
+    imports LESSON-22-SPEC-SYNTAX
+    imports LESSON-22
+
+endmodule
+
+module LESSON-22-SPEC
+    imports VERIFICATION
+
+endmodule
+```
+
+## 3. Claims
+
+1. The first claim we will ask K to prove is that 3 + 4, in fact, equals 7.
+   Claims are stated using the `claim` keyword, followed by the claim
+   statement:
+
+```
+claim <k> 3 + 4 => 7 ... </k>
+```
+
+Add this claim to the `LESSON-22-SPEC` module and run the K prover using the
+command `kprove lesson-22-spec.k`. You should get back the output `#Top`,
+which denotes the Matching Logic equivalent of `true` and means, in this
+context, that all claims have been proven correctly.
+
+2. The second claim reasons about the `if` statement that has a concrete condition:
+
+```
+claim <k> if ( 3 + 4 == 7 ) {
+            $a = 1 ;
+            } else {
+            $a = 2 ;
+            }
+        => . ... </k>
+        <store> STORE => STORE [ $a <- 1 ] </store>
+```
+
+stating that the given program terminates (`=> .`), and when it does, the value
+of the variable `$a` is set to `1`, meaning that the execution will have taken
+the `then` branch. Add this claim to the `LESSON-22-SPEC` module, but also add
+
+```
+syntax Id ::= "$a" [token]
+```
+
+to the `LESSON-22-SPEC-SYNTAX` module in order to declare `$a` as a token so
+that it can be used as a program variable. Re-run the K prover, which should
+again return `#Top`.
+
+3. Our third claim demonstrates how to reason about both branches of an `if`
+   statement at the same time:
+
+```
+claim <k> $a = A:Int ; $b = B:Int ;
+          if (A < B) {
+            $c = $b ;
+          } else {
+            $c = $a ;
+          }
+        => . ... </k>
+        <store> STORE => STORE [ $a <- A ] [ $b <- B ] [ $c <- ?C:Int ] </store>
+    ensures (?C ==Int A) orBool (?C ==Int B)
+```
```

Comment:
Would it make more sense to write `if ($a < $b) {` instead of `if (A < B) {`? My understanding is that this would make it a claim about programs that declare two variables and then compare them, etc. while currently it is a claim about programs which declare two variables `$a` and `$b` and then compares two `Int` values which happened to evaluate to the same values as `$a` and `$b` were initially assigned to, and the former behavior intuitively makes more sense to me as a property we would want to prove.